File: algo-visualizer/algorithms/searching.py
from typing import List, Tuple, Dict, Any
import math

SEARCH_META = {
    'linear': {'name':'Linear Search','tc':'O(n)','sc':'O(1)'},
    'binary': {'name':'Binary Search','tc':'O(log n)','sc':'O(1)','requires':'sorted array'},
    'jump': {'name':'Jump Search','tc':'O(sqrt(n))','sc':'O(1)','requires':'sorted array'},
    'interpolation': {'name':'Interpolation Search','tc':'O(log log n) average','sc':'O(1)','requires':'uniformly distributed sorted array'}
}

def generate_search_steps(arr: List[int], algorithm: str, target: int) -> Tuple[List[Dict], int]:
    steps=[]
    n=len(arr)
    if algorithm=='linear':
        for i,v in enumerate(arr):
            steps.append({'compare':[i,target]})
            if v==target:
                steps.append({'found':i}); return steps, i
        return steps, -1

    if algorithm=='binary':
        l=0; r=n-1
        while l<=r:
            m=(l+r)//2
            steps.append({'compare':[m,target]})
            if arr[m]==target: steps.append({'found':m}); return steps,m
            if arr[m] < target: l=m+1
            else: r=m-1
        return steps, -1

    if algorithm=='jump':
        step = int(math.sqrt(n))
        prev=0
        while prev < n and arr[min(n-1, prev+step-1)] < target:
            steps.append({'compare':[min(n-1, prev+step-1), target]})
            prev += step
        for i in range(prev, min(prev+step, n)):
            steps.append({'compare':[i,target]})
            if arr[i]==target: steps.append({'found':i}); return steps,i
        return steps, -1

    if algorithm=='interpolation':
        lo=0; hi=n-1
        while lo<=hi and target>=arr[lo] and target<=arr[hi]:
            if arr[hi]==arr[lo]:
                if arr[lo]==target: steps.append({'found':lo}); return steps, lo
                break
            pos = lo + int((target - arr[lo])*(hi-lo)/(arr[hi]-arr[lo]))
            steps.append({'compare':[pos, target]})
            if arr[pos]==target: steps.append({'found':pos}); return steps,pos
            if arr[pos] < target: lo = pos+1
            else: hi = pos-1
        return steps, -1

    return [], -1

File: algo-visualizer/static/js/searching.js
// Searching front-end
const barsEl = document.getElementById('bars');
const algoSel = document.getElementById('algo');
const arrayMode = document.getElementById('array-mode');
const customArray = document.getElementById('custom-array');
const btnGen = document.getElementById('generate');
const btnRun = document.getElementById('run');
const targetEl = document.getElementById('target');
const metaEl = document.getElementById('meta');
const resultEl = document.getElementById('result');

let arr = [];

function renderBars(){ barsEl.innerHTML=''; const w = Math.max(6, Math.floor(barsEl.clientWidth / arr.length) - 4);
  arr.forEach((h,i)=>{ const b=document.createElement('div'); b.className='bar'; b.style.height=(h*2)+'px'; b.style.width=w+'px'; b.dataset.index=i; b.title = h; barsEl.appendChild(b); }); }

btnGen.onclick = ()=>{ if(arrayMode.value==='custom' && customArray.value.trim()){ arr = customArray.value.split(',').map(x=>Number(x.trim())).filter(x=>!Number.isNaN(x)); } else if(arrayMode.value==='preset'){ arr = [1,3,5,7,9,11,13]; } else { arr = []; for(let i=0;i<20;i++) arr.push(1+Math.floor(Math.random()*100)); } if(['binary','jump','interpolation'].includes(algoSel.value)) arr.sort((a,b)=>a-b); renderBars(); metaEl.innerText=''; resultEl.innerText=''; }

btnRun.onclick = async ()=>{ const target = Number(targetEl.value); const res = await fetch('/api/search',{method:'POST', headers:{'Content-Type':'application/json'}, body: JSON.stringify({algorithm: algoSel.value, array: arr, target})}); const data = await res.json(); metaEl.innerText = `${data.meta.name} • TC/SC: ${data.meta.tc} / ${data.meta.sc}`; const steps = data.steps || []; for(const s of steps){ if(s.compare){ const i = s.compare[0]; const el = barsEl.querySelector(`[data-index="${i}"]`); if(el){ el.classList.add('compare'); await new Promise(r=>setTimeout(r,120)); el.classList.remove('compare'); } } if(s.found){ resultEl.innerText = 'Found at index: '+s.found; return; } } resultEl.innerText = 'Not found'; }

window.addEventListener('resize', renderBars);